### PR TITLE
topology2: mtl: Add HDMI-in capture support for SSP-based nocodec products.

### PR DESCRIPTION
**Pull request by CBala21**
Adding support for the product which doesn't have no ssp codec but need to support HDMI audio playback and HDMI-in capture via I2S.

### CODE REVIEW

**Review by lgirdwood — APPROVED** [review]

File: tools/topology/topology2/cavs-hdmi-in-ssp.conf [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	PLATFORM 			"none"
+}
+
+# override defaults with platform-specific config
+IncludeByKey.PLATFORM {
+	"mtl"	"platform/intel/mtl.conf"
+}
+
+# include HDMI-in capture config.
+IncludeByKey.HDMI_IN_CAPTURE {
```
Comment (ranj063) [inline tools/topology/topology2/cavs-hdmi-in-ssp.conf:69]: do we really need a new file for this? Can we not use sof-hda-generic and just add this IncludeByKey block alone? and set the BE link ID's in the tplg_targets.cmake file to generate the new tplg?

**Review by plbossart — CHANGES_REQUESTED** [review]
@CBala21 we've historically *never* enabled a combination of HDaudio + SSP. I don't know if this even works from a pinmux perspective, and I absolutely don't want to support this in the HDaudio machine driver.

NAK for now to make sure this configuration isn't merged.

File: tools/topology/topology2/cavs-hdmi-in-ssp.conf (hunk 1 above)
Comment (CBala21) [reply tools/topology/topology2/cavs-hdmi-in-ssp.conf:69]: Hi @ranj063,
 
No plan to add this No codec + HDMI-in capture support with HD audio. same time don't want to mix with any other I2S machine driver(ex: Es8326/RT5682). Since this no SSP codec+HDMI-In capture+HDMI audio playback is aligned with sof_ssp_amp machine driver and reuses the same for MTL. Also the plan is going forward for topology2 supported platforms will use this file to support no SSP codec+ HDMI-in capture via I2S + HDMI audio playback by using sof_ssp_amp.c machine driver.
Basically the intention is not to disturb/reduce complexity on I2S codec machine drivers and use the sof_ssp_amp.c machine driver for this use case.

**Comment by CBala21** [comment]
> @CBala21 we've historically _never_ enabled a combination of HDaudio + SSP. I don't know if this even works from a pinmux perspective, and I absolutely don't want to support this in the HDaudio machine driver.
> 
> NAK for now to make sure this configuration isn't merged.

Hi @plbossart ,

No plan to add this No codec + HDMI-in capture support with HD audio. Same time don't want to mix with any other I2S machine driver(ex: Es8326/RT5682). Since this no SSP codec+HDMI-In capture+HDMI audio playback is aligned with sof_ssp_amp machine driver (which has been used for ADL,RPL) and reuses the same for MTL. 
Also the plan is going forward for topology2 supported platforms will use this file to support no SSP codec+ HDMI-in capture via I2S + HDMI audio playback by using sof_ssp_amp.c machine driver.
Basically the intention is not to disturb and reduce complexity on I2S codec machine drivers and use the sof_ssp_amp.c machine driver for this use case.

**Comment by plbossart** [comment]
I am completely lost @CBala21. What exactly is a "non SSP codec"?

It seems that you are referring to a SSP-based nocodec solution, iow you have the negation on the wrong term?

**Comment by CBala21** [comment]
> I am completely lost @CBala21. What exactly is a "non SSP codec"?
> 
> It seems that you are referring to a SSP-based nocodec solution, iow you have the negation on the wrong term?

Hi @plbossart ,

yes. The product doesn't have any audio codec but need HDMI-In capture via I2S and HDMI audio playback. So we need I2S machine driver to support HDMI-In capture via I2S also HDMI audio playback. We are using sof-ssp_amp.c machine driver with I2S audio codec excluded from topology.

In topology1 we had a separate topology file for this sof_ssp_amp machine driver and we were reused it for ADL, RPL platforms to meet this use case.
But in topology2 we don't have topology file. hence I'm creating the new one without disturbing any other I2S machine drivers to meet the customer product use case.

**Comment by plbossart** [comment]
> > I am completely lost @CBala21. What exactly is a "non SSP codec"?
> > It seems that you are referring to a SSP-based nocodec solution, iow you have the negation on the wrong term?
> 
> Hi @plbossart ,
> 
> yes. The product doesn't have any audio codec but need HDMI-In capture via I2S and HDMI audio playback. So we need I2S machine driver to support HDMI-In capture via I2S also HDMI audio playback. We are using sof-ssp_amp.c machine driver with I2S audio codec excluded from topology.
> 
> In topology1 we had a separate topology file for this sof_ssp_amp machine driver and we were reused it for ADL, RPL platforms to meet this use case. But in topology2 we don't have topology file. hence I'm creating the new one without disturbing any other I2S machine drivers to meet the customer product use case.
> 
> Below is the quirk that we need to support from sof_ssp_amp.c and topology i have created to meet this case, { .name = "mtl_lt6911_hdmi_ssp", .driver_data = (kernel_ulong_t)(SOF_HDMI_CAPTURE_SSP_MASK(0x5) | /* SSP 0 and SSP 2 are used for HDMI IN */ SOF_NO_OF_HDMI_PLAYBACK(3) | SOF_HDMI_PLAYBACK_PRESENT), },

no following "We are using sof-ssp_amp.c machine driver with I2S audio codec excluded from topology." since there is nothing related to audio codec in the topology.

**Comment by CBala21** [comment]
> > > I am completely lost @CBala21. What exactly is a "non SSP codec"?
> > > It seems that you are referring to a SSP-based nocodec solution, iow you have the negation on the wrong term?
> > 
> > 
> > Hi @plbossart ,
> > yes. The product doesn't have any audio codec but need HDMI-In capture via I2S and HDMI audio playback. So we need I2S machine driver to support HDMI-In capture via I2S also HDMI audio playback. We are using sof-ssp_amp.c machine driver with I2S audio codec excluded from topology.
> > In topology1 we had a separate topology file for this sof_ssp_amp machine driver and we were reused it for ADL, RPL platforms to meet this use case. But in topology2 we don't have topology file. hence I'm creating the new one without disturbing any other I2S machine drivers to meet the customer product use case.
> > Below is the quirk that we need to support from sof_ssp_amp.c and topology i have created to meet this case, { .name = "mtl_lt6911_hdmi_ssp", .driver_data = (kernel_ulong_t)(SOF_HDMI_CAPTURE_SSP_MASK(0x5) | /* SSP 0 and SSP 2 are used for HDMI IN */ SOF_NO_OF_HDMI_PLAYBACK(3) | SOF_HDMI_PLAYBACK_PRESENT), },
> 
> no following "We are using sof-ssp_amp.c machine driver with I2S audio codec excluded from topology." since there is nothing related to audio codec in the topology.

> Sorry for the confusion, for this use case sof-ssp_amp.c machine driver has been used and not adding the I2S config for headset/speaker codec in the topology file to make it clear that topology file not supported I2S codec config and supporting only HDMI-in capture I2S config and HDMI configs.

**Comment by plbossart** [comment]
I think you're taking the problem from the wrong direction.

Everything you do is based on sof_ssp_amp.c, i.e. something that does not include a headset codec.

The first order of business should be to add such a topology configuration in a modular way, and second to add the HDMI input also with a conditional include. It makes no sense to me to create a configuration that can only be used on your platform, I mean, from your perspective it's probably a shortcut but it's more maintenance for the rest of us.

**Comment by CBala21** [comment]
Based on @plbossart comment, dropping the changes and uploaded the suggested approach in new PR.
https://github.com/thesofproject/sof/pull/8784
